physical/couchdb.go

```diff
@@ -0,0 +1,285 @@
+	}
+
+	return &CouchDBBackend{
+		client: &couchDBClient{
+			endpoint: endpoint,
+			username: username,
+			password: password,
+			Client:   &http.Client{},
+		},
+		logger:     logger,
+		permitPool: NewPermitPool(DefaultParallelOperations),
+	}, nil
+}
+
+func newCouchDBBackend(conf map[string]string, logger log.Logger) (Backend, error) {
+	return buildCouchDBBackend(conf, logger)
+}
+
+type couchDBEntry struct {
+	Entry   *Entry `json:"entry"`
+	Rev     string `json:"_rev,omitempty"`
+	ID      string `json:"_id"`
+	Deleted *bool  `json:"_deleted,omitempty"`
+}
+
+// Put is used to insert or update an entry
+func (m *CouchDBBackend) Put(entry *Entry) error {
+	m.permitPool.Acquire()
+	defer m.permitPool.Release()
+
+	defer metrics.MeasureSince([]string{"couchdb", "put"}, time.Now())
+
+	revision, _ := m.client.rev(url.PathEscape(entry.Key))
+
+	return m.client.put(couchDBEntry{
+		Entry: entry,
+		Rev:   revision,
+		ID:    url.PathEscape(entry.Key),
+	})
+}
+
+// Get is used to fetch an entry
+func (m *CouchDBBackend) Get(key string) (*Entry, error) {
+	m.permitPool.Acquire()
+	defer m.permitPool.Release()
+
+	defer metrics.MeasureSince([]string{"couchdb", "get"}, time.Now())
+
+	return m.client.get(key)
+}
+
+// Delete is used to permanently delete an entry
+func (m *CouchDBBackend) Delete(key string) error {
+	m.permitPool.Acquire()
+	defer m.permitPool.Release()
+
+	defer metrics.MeasureSince([]string{"couchdb", "delete"}, time.Now())
+
+	revision, _ := m.client.rev(url.PathEscape(key))
+	deleted := true
+	return m.client.put(couchDBEntry{
+		ID:      url.PathEscape(key),
+		Rev:     revision,
+		Deleted: &deleted,
+	})
+}
+
+// List is used to list all the keys under a given prefix
+func (m *CouchDBBackend) List(prefix string) ([]string, error) {
+	defer metrics.MeasureSince([]string{"couchdb", "list"}, time.Now())
+
+	items, err := m.client.list(prefix)
+	if err != nil {
+		return nil, err
+	}
+
+	var out []string
+	seen := make(map[string]interface{})
+	for _, result := range items {
+		trimmed := strings.TrimPrefix(result.ID, prefix)
+		sep := strings.Index(trimmed, "/")
+		if sep == -1 {
+			out = append(out, trimmed)
+		} else {
+			trimmed = trimmed[:sep+1]
+			if _, ok := seen[trimmed]; !ok {
+				out = append(out, trimmed)
+				seen[trimmed] = struct{}{}
+			}
+		}
+	}
+	return out, nil
+}
+
+// TransactionalCouchDBBackend creates a couchdb backend that forces all operations to happen
+// in serial
+type TransactionalCouchDBBackend struct {
+	CouchDBBackend
+}
+
+func newTransactionalCouchDBBackend(conf map[string]string, logger log.Logger) (Backend, error) {
+	backend, err := buildCouchDBBackend(conf, logger)
+	if err != nil {
+		return nil, err
+	}
+	backend.permitPool = NewPermitPool(1)
+
+	return &TransactionalCouchDBBackend{
+		CouchDBBackend: *backend,
+	}, nil
+}
+
+// GetInternal is used to fetch an entry
+func (m *TransactionalCouchDBBackend) GetInternal(key string) (*Entry, error) {
+	return m.Get(key)
```

Comment:
This is backwards; the entire point of the 'internal' functions are that they *don't* acquire a permit.

Rather than GetInternal chaining to Get, Get should chain to GetInternal where the heavy lifting happens. Both Get and GetInternal should be methods of the base CouchDBBackend.